### PR TITLE
Owensmallwood/add gh action to build and push docker image

### PR DESCRIPTION
**Pull request by owensmallwood**
**Notes**
Leaving this in draft until we have a secret for Dockerhub.

**Changes**
This will build, tag, and push a new docker image to Dockerhub when a new release is published.

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+name: Build and Push Docker image
+
+on:
+  release:
+    types: [published]
```
Comment (hairyhenderson) [inline .github/workflows/release.yaml:5]: I'd like to build on all PRs, but only push on merges to main - I think this means it'll  only build when releases are published?

Comment (hairyhenderson) [reply .github/workflows/release.yaml:5]: oh, actually we already do build on PRs - https://github.com/grafana/smtprelay/blob/1a6e558df42ce7c96ffaa960d8b1a4b606592a5c/.github/workflows/docker.yml

so maybe this is OK... 🤔

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+          images: grafana/smtprelay
+
+      - name: Build and push Docker image
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          file: ../../Dockerfile
+          push: true
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
```
Comment (hairyhenderson) [inline .github/workflows/release.yaml:34]: I'm curious what these will include - I wonder if we can run this with `push: false` in this PR just to see the logs?

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          images: grafana/smtprelay
+
+      - name: Build and push Docker image
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          file: ../../Dockerfile
```
Comment (hairyhenderson) [inline .github/workflows/release.yaml:31]: I think these can be removed

```suggestion
```

Comment (owensmallwood) [reply .github/workflows/release.yaml:31]: ah yeah, file's default is `{context}/Dockerfile` and the default context should be repo root then. cool 👍

File: .github/workflows/release.yaml (hunk 2 above)
Comment (owensmallwood) [reply .github/workflows/release.yaml:32]: I'm also unsure. Will comment out the push for now

File: .github/workflows/release.yaml (hunk 1 above)
Comment (owensmallwood) [reply .github/workflows/release.yaml:5]: I guess it should tag and push the image when we do a new release, and if we push to main it should build and push with the `latest` tag

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
+name: Build and Push Docker image
+
+on:
+  push:
+    branches:
+      - main
```
Comment (hairyhenderson) [inline .github/workflows/release.yaml:6]: I think we can test it out on this branch with the following:

```suggestion
    branches:
      - main
      - owensmallwood/add-gh-action-to-build-and-push-docker-image
```

Comment (owensmallwood) [reply .github/workflows/release.yaml:6]: ah, good call 👍

**Review by hairyhenderson — APPROVED** [review]
LGTM - let's merge this and tweak from there if we need to.

**Pull request merged**